_ext/eclipse-wtp/src/main/java/com/diffplug/spotless/extra/eclipse/wtp/sse/ContentTypeManager.java

```diff
@@ -83,6 +86,45 @@ public IContentType findContentTypeFor(InputStream contents, String fileName) th
 		return processorStepType;
 	}
 
+	@Override
+	public IContentDescription getDescriptionFor(InputStream contents, String fileName, QualifiedName[] options) throws IOException {
+		return processorStepDescription;
+	}
+
+	private static class StringDescription implements IContentDescription {
+
+		private final IContentType type;
+
+		public StringDescription(IContentType type) {
+			this.type = type;
+		}
+
+		@Override
+		public boolean isRequested(QualifiedName key) {
+			return false; //Don't use set Property
+		}
+
+		@Override
+		public String getCharset() {
+			return Charset.defaultCharset().name(); //Spotless operates on an decoded string, meaning the input has always the "internal" encoding
```

Comment:
I think this should be `UTF-8`, or any other `UTF`, but `Charset.defaultCharset()` will cause a problem if the machine happens to have a non-unicode default encoding.  For example, if I were in Japan, my default charset might be `Shift_JIS`, and if I tell Spotless to read a file with `UTF-8`, Spotless will correctly load the binary with`UTF`, and the string may include codepoints which are *not encodable* in `Shift_JIS`.  In #545, I bet their machine is using `Latin-1` as the default charset, and the mangled codepoints are not encodable in `Latin-1`.

The nice thing about unicode is that, no matter which unicode you pick, every codepoint is encodable, so you always get a lossless round trip.  If you rely on the default, you might get a default which cannot encode the full unicode plane.

Reply:
I am afraid I disagree. You are right, my current solution is not correct.
But UTF-8 and UTF-16 are completely different on bit level. They can be converted, yes, but you cannot use a UFT-16 reader and give it a UTF-8 stream. The result would be garbage.
That's why I don't like that Sptoless keeps BOM in the stream when reading a file. You end up for example with a UTF-8 BOM, but the string (since native encoded) has a UTF-16 encoding.

Reply:
I hope my new comments in the code are convincing. Otherwise please reopen this conversation.

Reply:
Character encodings have two parts - the total set of representable characters (e.g. unicode, but there used to be others) and the binary encoding used for those characters.  A java.lang.String doesn't *have* an encoding, sure there is a UTF-16 encoded interface in the `charAt` methods, and also a full codepoint interface in the `codePoints` method, but the "encoding" is an opaque implementation detail - might be UTF-16 in a char array, might be UTF-8 in a byte[] array, someday it might be UTF-32 in an int array for Chinese-optimized JVMs.

Because we are passing a String to eclipse, there is no chance that we might use the wrong "bit encoding" - eclipse is only able to get those bytes by encoding the String, so you always get a sensible result whether you use `UTF-8`, `UTF-16`, `UTF-32`, or even `UTF-16BE`.  However, if you don't use *some* UTF, then there will be parts of the unicode character set which are not representable.  That's why any UTF will work, and every non-UTF will not work.